pkg/volumes/alicloud/volumes.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	config := alicloud.NewConfig()
+
+	credential := credentials.NewEcsRamRoleCredential(ramRole)
+	a.ecs, err = ecs.NewClientWithOptions(a.region, config, credential)
+
+	return a, nil
+}
+
+// AttachVolume attaches the specified volume to this instance, returning the mountpoint & nil if successful
+func (a *AlicloudVolumes) AttachVolume(volume *volumes.Volume) error {
+	volumeID := volume.ProviderID
+
+	if volume.LocalDevice == "" {
+		attachDiskReq := ecs.CreateAttachDiskRequest()
+		attachDiskReq.RegionId = a.region
+		attachDiskReq.InstanceId = a.instanceID
+		attachDiskReq.DiskId = volumeID
+
+		attachResponse, err := a.ecs.AttachDisk(attachDiskReq)
+		if err != nil {
+			return fmt.Errorf("Unable to attach volume %q: %v", volumeID, err)
+		}
+
+		glog.V(2).Infof("AttachVolume request returned %v", attachResponse)
+
+		// Attaching volume takes time.
+		time.Sleep(5)
+	}
+
+	for {
+		request := ecs.CreateDescribeDisksRequest()
+		request.PageSize = requests.NewInteger(PageSizeLarge)
+		request.PageNumber = requests.NewInteger(1)
+		request.DiskIds = fmt.Sprintf("[\"%s\"]", volumeID)
+
+		volumes, err := a.describeVolumes(request)
+		if err != nil {
+			return fmt.Errorf("Error describing volume %q: %v", volumeID, err)
+		}
+
+		if len(volumes) == 0 {
+			return fmt.Errorf("Disk volume %q disappeared during attach", volumeID)
+		}
+		if len(volumes) != 1 {
+			return fmt.Errorf("Multiple volumes found with id %q", volumeID)
+		}
+
+		v := volumes[0]
+		if v.AttachedTo != "" {
+			if v.AttachedTo == a.instanceID {
+				volume.LocalDevice = v.LocalDevice
+				return nil
+			} else {
+				return fmt.Errorf("Unable to attach volume %q, was attached to %q", volumeID, v.AttachedTo)
+			}
+		}
+
+		switch v.Status {
+		case "Attaching":
+			glog.V(2).Infof("Waiting for volume %q to be attached (currently %q)", volumeID, v.Status)
+			// continue looping
+
+		default:
+			return fmt.Errorf("Observed unexpected volume state %q", v.Status)
+		}
+
+		time.Sleep(10 * time.Second)
+	}
+}
+
+func (a *AlicloudVolumes) FindVolumes() ([]*volumes.Volume, error) {
+	return a.findVolumes(true)
+}
+
+func (a *AlicloudVolumes) findVolumes(filterByAZ bool) ([]*volumes.Volume, error) {
+	request := ecs.CreateDescribeDisksRequest()
+	request.RegionId = a.region
+
+	if filterByAZ {
+		request.ZoneId = a.az
+	}
+
+	var tags []ecs.DescribeDisksTag
+
+	for _, key := range a.matchTagKeys {
+		tags = append(tags, ecs.DescribeDisksTag{
+			Key:   key,
+			Value: "",
+		})
+	}
+
+	for key, value := range a.matchTags {
+		tags = append(tags, ecs.DescribeDisksTag{
+			Key:   key,
+			Value: value,
+		})
+	}
+
+	request.Tag = &tags
+	request.PageSize = requests.NewInteger(PageSizeLarge)
+	request.PageNumber = requests.NewInteger(1)
+
+	return a.describeVolumes(request)
+}
+
+func (a *AlicloudVolumes) describeVolumes(request *ecs.DescribeDisksRequest) ([]*volumes.Volume, error) {
+	var found []*volumes.Volume
+
+	for {
```

Comment:
I am not big fan of endless loops, if you want keep this logic, maybe we should put log to each loop? For instance "describing volumes page x". Then we can easily see if it is stuck. This can potentially make requests quite heavily against cloud apis.

Reply:
@zetaab I couldn't think out a better way to do the describing without a endless for loop here.

So I added more logging and rebased.